### PR TITLE
VAD for browser client beta

### PR DESCRIPTION
**Pull request by arzga**
### What

- Introduced energy threshold VAD akin to Unity implementation. Code resides in `audioprocessor folder`, and is running in the web worker that also handles websocket connectivity.
- Added start/stopStream to control feeding audio constantly to audio processor, currently only used internally. StopContext ensures all data in internal buffer is processed.
- Added `rollup-plugin-web-worker-loader` to build tooling to create web worker from multiple source files

#### How the VAD works

- VAD normally signals to CloudDecoder when signal has been detected so it can control start / stop context
- VAD docs
- There is a special immediate mode used by file upload, indicated by `immediate` flag in startStream. This enables the worker to control start / stop context autonomously (Note: this might be changed in the future in favour of bidirectional  audioprocessor control to slow down data streaming to level accepted by the backend)

### Why

- VAD can enable hands free use in some use cases
- VAD can save network bandwidth and create cleaner utterances

### CODE REVIEW

**Review by samutamm — COMMENTED** [review]
Looks good! I just left few minor nit-picks...

File: examples/browser-client-example/src/index.ts [hunk 1]
```diff
@@ -69,18 +73,20 @@ window.onload = () => {
 };
 
 function newDecoder(): CloudDecoder {
+  // const appId = "d9abea67-18e5-4c4e-b7fc-51f66d3219e2";
   const appId = process.env.REACT_APP_APP_ID || "be3bfb17-ee36-4050-8830-743aa85065ab";
```
[the remark this answers is not in the document]
Comment (arzga) [reply examples/browser-client-example/src/index.ts]: Removed my staging test id

File: libraries/browser-client/src/websocket/worker.ts [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+      this.audioProcessor.onVadSignalHigh = () => {
+        const currentVadOptions = this.audioProcessor?.vad?.vadOptions
+        if (currentVadOptions) {
+          if (this.defaultContextOptions?.immediate) {
+            if (currentVadOptions.enabled && currentVadOptions.controlListening) {
+              this.startContext()
+            }
+          } else {
+            if (this.audioProcessor?.vad?.vadOptions.enabled) {
```
Comment (samutamm) [inline libraries/browser-client/src/websocket/worker.ts:91]: At the line 92 `currentVadOptions.enabled` verifies the same thing, this line could be removed

File: libraries/browser-client/src/websocket/worker.ts [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+  initAudioProcessor(sourceSampleRate: number, vadOptions?: VadOptions): void {
+    this.audioProcessor = new AudioProcessor(sourceSampleRate, this.targetSampleRate, 5)
+
+    if (vadOptions) {
+      this.audioProcessor.vad = new EnergyTresholdVAD(vadOptions)
+
+      this.audioProcessor.onVadSignalHigh = () => {
+        const currentVadOptions = this.audioProcessor?.vad?.vadOptions
+        if (currentVadOptions) {
```
Comment (samutamm) [inline libraries/browser-client/src/websocket/worker.ts:85]: 5 levels of nesting is a bit hard to read, I suggest to return right away if `currentVadOptions` are not defined
```suggestion
        if (currentVadOptions === undefined) return
```

File: libraries/browser-client/src/websocket/worker.ts [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+              if (currentVadOptions.enabled && currentVadOptions.controlListening) {
+                this.workerCtx.postMessage({ type: WorkerSignal.VadSignalHigh })
+              }
+            }
+          }
+        }
+      }
+      this.audioProcessor.onVadSignalLow = () => {
+        const currentVadOptions = this.audioProcessor?.vad?.vadOptions
+        if (currentVadOptions) {
```
Comment (samutamm) [inline libraries/browser-client/src/websocket/worker.ts:101]: For less nesting, return right away
```suggestion
        if (currentVadOptions === undefined) return
```

**Review by samutamm — APPROVED** [review]

File: libraries/browser-client/src/websocket/worker.ts (hunk 3 above)
Comment (arzga) [reply libraries/browser-client/src/websocket/worker.ts:85]: Thanks for pointing this out. Did some cleanup there and it's much better now.

**Pull request merged**